urdf_usd_converter/_impl/link.py

```diff
@@ -1,10 +1,295 @@
+                link = self.get_link_by_name(joint.child.get_with_default("link"))
+                self.link_tree[parent_link_name]["children"].append(link)
+                self.link_tree[parent_link_name]["joints"].append(joint)
+
+        # If the link tree is empty, make the first link the root.
+        if len(self.link_tree) == 0:
+            link = self.root_element.links[0]
+            self.link_tree[link.name] = {
+                "link": link,
+                "joints": [],
+                "children": [],
+            }
+
+    def get_root_link(self) -> ElementLink:
+        """
+        Get the root link name from the link hierarchy.
+        """
+        links = [data["link"] for data in self.link_tree.values()]
+        for link in links:
+            is_child = False
+            for d in self.link_tree.values():
+                if link in d["children"]:
+                    is_child = True
+                    break
+            if not is_child:
+                return link
+
+        # If it is a looping joint structure, the process reaches this point.
+        raise ValueError("No root link found. The joint structure is a loop.")
+
+    def get_link_joints(self, link_name: str) -> list[ElementJoint]:
+        """
+        Get the joints that connect to a link.
+        """
+        if link_name not in self.link_tree:
+            return None
+        return self.link_tree[link_name]["joints"]
+
+    def get_link_children(self, link_name: str) -> list[ElementLink]:
+        """
+        Get the children of a link.
+        """
+        if link_name not in self.link_tree:
+            return []
+        return self.link_tree[link_name]["children"]
+
+    def get_link_by_name(self, link_name: str) -> ElementLink:
+        """
+        Get a link by name.
+        """
+        return next((link for link in self.root_element.links if link.name == link_name), None)
+
+
+def convert_link(parent: Usd.Prim, link_hierarchy: LinkHierarchy, link: ElementLink, data: ConversionData) -> UsdGeom.Xform:
+    link_safe_name = data.name_cache.getPrimName(parent, link.name)
+    link_xform = usdex.core.defineXform(parent, link_safe_name)
+    link_prim = link_xform.GetPrim()
+    if link.name != link_safe_name:
+        usdex.core.setDisplayName(link_prim, link.name)
+
+    data.references[Tokens.Physics][link.name] = link_prim
+
+    apply_physics_rigidbody(link_prim, data)
+
+    # Assigning MassAPI to a Rigid Body.
+    apply_inertial(link_prim, link, data)
+
+    # Create visual or collision geometry.
+    geometry_basses: list[ElementVisual | ElementCollision] = [visual for visual in link.visuals if visual.geometry and visual.geometry.shape] + [
+        collision for collision in link.collisions if collision.geometry and collision.geometry.shape
+    ]
+
+    for geometry_base in geometry_basses:
+        name = get_geometry_name(geometry_base)
+        geom_safe_name = data.name_cache.getPrimName(link_prim, name)
```

Comment:
you should get all the safe names at once outside the loop, it is more efficient